src/Compiler/Microsoft.AspNetCore.Razor.Language/src/BuilderExtensions.cs

```diff
@@ -0,0 +1,24 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Razor.PooledObjects;
+using Microsoft.Extensions.ObjectPool;
+
+namespace Microsoft.AspNetCore.Razor.Language;
+
+internal static class BuilderExtensions
+{
+    public static T[] BuildAll<T, TBuilder>(this List<TBuilder> builders, ObjectPool<HashSet<T>> hashSetPool)
+        where TBuilder : IBuilder<T>
+    {
+        using var _ = hashSetPool.GetPooledObject(out var set);
```

Comment:
❓ is there a `GetPooledObject` that takes the required minimum size?

Reply:
Not yet, I'll add that.

Reply:
Unfortunately, `HashSet<T>` does not have a facility on .NET Framework or .NET Standard 2.0 for setting the capacity after it has been created.

Reply:
On .NET Core and .NET Standard 2.1, `HashSet<T>.EnsureCapacity(...)` is available, but we can't use that in VS. 😞